File: Namyata/CheckIfMonotonicArray.py
import unittest
from typing import List
# An array is monotonic if it is either monotone increasing or monotone decreasing.
#
# An array nums is monotone increasing if for all i <= j, nums[i] <= nums[j]. An array nums is monotone decreasing if for all i <= j, nums[i] >= nums[j].
#
# Given an integer array nums, return true if the given array is monotonic, or false otherwise.



def isMonotonic(nums: List[int]) -> bool:
    return nums == sorted(nums) or nums == sorted(nums, reverse=True)

class TestIsMonotonic(unittest.TestCase):

    def test_increasing_list(self):
        nums = [1, 2, 3, 4, 5]
        expected = True
        self.assertEqual(isMonotonic(nums), expected)

    def test_decreasing_list(self):
        nums = [5, 4, 3, 2, 1]
        expected = True
        self.assertEqual(isMonotonic(nums), expected)

    def test_monotonic_list(self):
        nums = [1, 3, 5, 7, 9]
        expected = True
        self.assertEqual(isMonotonic(nums), expected)

    def test_non_monotonic_list(self):
        nums = [1, 3, 2, 5, 4]
        expected = False
        self.assertEqual(isMonotonic(nums), expected)

    def test_equal_elements_list(self):
        nums = [5, 5, 5, 5, 5]
        expected = True
        self.assertEqual(isMonotonic(nums), expected)

if __name__ == '__main__':
    unittest.main()

#All unit tests are correct and passed

File: CTests/check_if_monotonic_array.c
#include <stdio.h>
#include <stdbool.h>
#include <assert.h>

bool isMonotonic(int* nums, int numsSize) {
    bool increasing = true;
    bool decreasing = true;

    for (int i = 1; i < numsSize; i++) {
        if (nums[i] < nums[i - 1]) {
            increasing = false;
        }
        if (nums[i] > nums[i - 1]) {
            decreasing = false;
        }
    }

    return increasing || decreasing;
}

void test_increasing_list() {
    int nums[] = {1, 2, 3, 4, 5};
    int numsSize = sizeof(nums) / sizeof(nums[0]);
    bool result = isMonotonic(nums, numsSize);
    bool expected = true;
    assert(result == expected);
    printf("Test case 'increasing_list' passed\n");
}

void test_decreasing_list() {
    int nums[] = {5, 4, 3, 2, 1};
    int numsSize = sizeof(nums) / sizeof(nums[0]);
    bool result = isMonotonic(nums, numsSize);
    bool expected = true;
    assert(result == expected);
    printf("Test case 'decreasing_list' passed\n");
}

void test_monotonic_list() {
    int nums[] = {1, 3, 5, 7, 9};
    int numsSize = sizeof(nums) / sizeof(nums[0]);
    bool result = isMonotonic(nums, numsSize);
    bool expected = true;
    assert(result == expected);
    printf("Test case 'monotonic_list' passed\n");
}

void test_non_monotonic_list() {
    int nums[] = {1, 3, 2, 5, 4};
    int numsSize = sizeof(nums) / sizeof(nums[0]);
    bool result = isMonotonic(nums, numsSize);
    bool expected = false;
    assert(result == expected);
    printf("Test case 'non_monotonic_list' passed\n");
}

void test_equal_elements_list() {
    int nums[] = {5, 5, 5, 5, 5};
    int numsSize = sizeof(nums) / sizeof(nums[0]);
    bool result = isMonotonic(nums, numsSize);
    bool expected = true;
    assert(result == expected);
    printf("Test case 'equal_elements_list' passed\n");
}

int main() {
    test_increasing_list();
    test_decreasing_list();
    test_monotonic_list();
    test_non_monotonic_list();
    test_equal_elements_list();

    return 0;
}
